### ISSUE TITLE
Take into account load activity on a node when routing search requests

### ISSUE THREAD

**Issue by ppf2** [elastic/elasticsearch#12496]
Currently, by default search routing across copies of a shard is done in a randomized fashion.  But it does not take into account the load of the node the shard copy is on.  For a high search throughput cluster where there are many replicas configured, if a node becomes loaded, it will be nice for the routing logic to be smarter and take load, search queue length/rejections, etc.. into consideration when routing requests

**Comment by nik9000** [comment]
This is a bit scary to me. Its already possible for a single node to poison the cluster. I'm afraid this trades some of those problems for more problems.

Imagine you have a cluster that consistently runs a queue depth of 20 searches. But something happens to one node causing it to fail all requests quickly. Now you start sending tons of requests to that node effectively killing the cluster. You can work around this by adding more smarts to detect the error.

I'm just worried this is a big thing to bite off and a difficult one to get right.

**Comment by bakks** [comment]
It only took 8 minutes to shoot this down, that must be some kind of record!

Here's some background on this problem which we're observing with our cluster. We have a 24 node cluster (16 cores on each), 2 indexes, 6 shards in each, replica factor of 3, so 24 shards in each index, and 2 shards per node. Here's a chart of load on the cluster members:
![screen shot 2015-07-27 at 3 03 29 pm](https://cloud.githubusercontent.com/assets/1172710/8919161/0826e89c-3472-11e5-9d38-1f2118bee49c.png)
When we look at the `_cat/thread_pool` output that node consistently has all of its search threads busy and a large number of queued requests.

We seem to have some as-of-yet undiagnosed pathological query problems, which means that our entire ES cluster is CPU bound. When we have problems its generally because CPU is capping on one or more nodes. As you'll observe from the chart above, a single node has much higher load than the other 23 nodes. This is the effect on application performance:
![screen shot 2015-07-27 at 3 17 19 pm](https://cloud.githubusercontent.com/assets/1172710/8919260/c37a50ac-3472-11e5-9b7e-95ca7cb805ce.png)

Effectively the fact that a single node is capping CPU means that the average latency for our search traffic goes from 20 ms to 1000+ ms. This is bad in itself, but also our entire application effectively must queue web requests because a core backend service is stressed because a _single node_ is having trouble. In my opinion this behavior obviates the entire point of running a distributed system, which is to divide the workload and handle isolated problems.

I'd like to request that this be looked at with a high priority - either there's something seriously broken about our deployment or this is a serious problem with ES query routing. I'm more than happy to provide more details about our deployment, @ppf2 has a diagnostic dump of what this looks like in practice. Thanks for looking at this!

**Comment by nik9000** [comment]
Something is certainly wrong. @ppf2 will certainly know more with the dump. I'm just weary of any kind of automatic actions on single nodes that are in trouble because it makes me think of cascading failures. I'm sure there are good, safe things that Elasticsearch could do but I worry about doing something simple without thinking it through very very very hard.

**Comment by bakks** [comment]
Here's the functionality I would suggest: pick shards for queries based on queue lengths. If a single node has a much larger queue length than every other node, it probably shouldn't be receiving more queries.

My understanding is that ES currently randomly selects among the running nodes that are serving the desired shard. So if a node drops out of the cluster, ES removes it from the pool and all is well. If a node is stressed, ie under higher load than its peers, then the entire cluster's performance is affected, because queries continue to be blindly routed to this node. As you add nodes to your cluster, the chances of a single node encountering issues increases.

I would invite you to test a large ES cluster and stress a single one of the nodes, observing query latency when this occurs.

I think the high-level question here is whether or not ES should be robust in situations where a node remains in the cluster but doesn't perform as well as other nodes. I would argue that its a poor distributed system if these scenarios are ignored. Currently its very easy for a single node to negatively impact overall ES query performance.

**Comment by clintongormley** [comment]
@bakks it is not a question of importance - this is clearly an important thing to get right.  the problem is the second part: getting it right.  Simple heuristics are quite likely to make things worse.  We are investigating more robust techniques for taking load into account, but it is not something that we are going to be able to rush in.  It is a big project which needs a lot of careful testing before we can make any changes.

In your case, I'd try to figure out why one node is taking the brunt of the load.  Are you

* sending all search requests to one node?
* performing lots of updates?

Would be interesting to get a hot threads dump for the node in particular, especially compared to another node that doesn't have so much CPU usage.

**Comment by bakks** [comment]
@clintongormley Thanks for your response - totally understand that you guys want to get this right.

What I was pointing out was not that we have a more loaded node than others, but rather that ES seems to handle that case very poorly. We distribute queries evenly among our data nodes and our updates are spread fairly evenly, so I don't think there's anything about our workload thats unbalanced in terms of ES shards. When we're running normally our nodes look similar in terms of observed load.

However, this ES characteristic is still relevant to us and others because:
- I think its pretty common to have a cluster spread over heterogeneous hardware. For example, In our cloud provider we split nodes into 3 separate zones, one of which runs an older generation CPU.
- Exogenous factors can effect node performance. Noisy neighbors, for example. In fact, as you increase the number of nodes the chance of an event outside your control affecting node performance also increases. Resiliency to node issues is obviously one of the key reasons to run a distributed system. Because ES doesn't handle single-node performance issues, you actually decrease your cluster stability when you add data nodes.
- As you approach the performance ceiling of a cluster, inevitably query queues increase and some node will be the first to overflow its queue and reject requests. From what I can tell, failing to route around this node effectively lowers Elasticsearch's query throughput ceiling.

@ppf2 has access to a bunch of telemetry from our clusters when we're experiencing heavy load, more than welcome to take a look at those as you investigate more.

**Comment by bakks** [comment]
Just a ping on this issue - it is a serious problem with Elasticsearch's stability during load. Any movement on it?

**Comment by ppf2** [comment]
Related: https://github.com/elastic/elasticsearch/issues/15914

**Comment by eskibars** [comment]
@clintongormley @dakrone with https://github.com/elastic/elasticsearch/issues/24915 is there anything else we want to explore here?

**Comment by dakrone** [comment]
@eskibars I believe this is resolved for now, there are more improvements and tests that can be made with regard to adaptive replica selection, but I will open a separate issue for those.
